crates/project/src/project.rs

```diff
@@ -3523,6 +3533,19 @@ impl Project {
         })
     }
 
+    pub fn folding_ranges(
```

Comment:
This ended quite unexpectedly on the most interesting part: how to use that??

We have to consider 2 things:

* whether those should be mixed along with the tree-sitter queries.
I suspect that should not happen, hence we need to do something like https://github.com/zed-industries/zed/blob/cb0e85cf993a0fd94a4e44956a11f27ccf11ff7a/assets/settings/default.json#L1461-L1470

* where, how to store all this data, how to invalidate and re-query this.
This is the gist of this PR and I'd expect that to be in a "ready for review" one.